### PR TITLE
Refactor repositories pool

### PR DESCRIPTION
**Pull request by b-kamphorst**
# Pull Request Check List

Relates-to: #3155

- [x] Added **tests** for changed code.
- [ ] Updated **documentation** for changed code.

I refactored `Pool` in preparation of tackling https://github.com/python-poetry/poetry/issues/6713. Most notably, this PR uses an enum to significantly reduce the bookkeeping effort for the various types of repositories (primary / default / secondary).

Two considerations:
- One may argue whether the `RepositoryPriority` (name open for suggestions) should be an attribute of a `Repository` instead. I'm willing to have a look at that, but as that change will scatter all over the (test) code I decided to present this suggestion first.
- `OrderedDict` is used just to satisfy the test `test_repository_ordering` in `tests/repositories/test_pool.py`. Although I appreciate the deterministic approach, I would like to verify whether this ordering is actually relevant. In particular, was the order only verified to ensure proper ordering + clustering between default/primary/secondary, or also proper ordering within each cluster? Phrased differently, does the user expect that the order in which primary sources A and B are added through the CLI affects the resolution order? They might and I'm happy to leave it as is, but otherwise we can use normal dicts.

### CODE REVIEW

**Comment by b-kamphorst** [comment]
I also removed the inheritance from `Repository` without removing any functionality. The only thing that `Pool` actually used was the `name` property. As it appears that this is not used anywhere, I made `Pool` nameless.

**Comment by radoering** [comment]
> Phrased differently, does the user expect that the order in which primary sources A and B are added through the CLI affects the resolution order?

I expect that the order of primary sources in `pyproject.toml` is relevant. I didn't find anything in the docs, but I think we should keep it deterministic.

**Comment by b-kamphorst** [comment]
Hi @radoering, thank you for your swift consideration.

> > Phrased differently, does the user expect that the order in which primary sources A and B are added through the CLI affects the resolution order?
> 
> I expect that the order of primary sources in `pyproject.toml` is relevant. I didn't find anything in the docs, but I think we should keep it deterministic.

That is very reasonable. I've clarified the resolution order in the docs. Please have a look at the proposed change. I'm also looking forward to your other thoughts on the PR.

**Comment by radoering** [comment]
I just ran `test_pool.py` with coverage. It seems there is no coverage for `has_primary_repositories`, `find_packages` and `search` and `package` is only tested superficially. It would be nice to add some tests to gain more confidence in the refactoring.

I still need to take a closer look, but at first glance it looks like a neat simplification. 👍

However, I noticed that without `_lookup` and the Boolean flags there is more iterating in some methods/properties. I don't know how often this code is called during dependency resolution and if this could be a performance bottleneck.

File: src/poetry/repositories/pool.py
Comment (neersighted) [inline src/poetry/repositories/pool.py:27]:
```suggestion
class RepositoryPool:
```

Let's go all the way with breaking the false inheritance and name this clearly.

File: src/poetry/repositories/pool.py
Comment (neersighted) [inline src/poetry/repositories/pool.py:90]: While we are refactoring, I think we should hoist any normalization up and out (or make it explicit that names are not normalized) -- having this only in the remove code is surprising.

**Review by neersighted — CHANGES_REQUESTED** [review]

File: src/poetry/repositories/pool.py
Comment (dimbleby) [reply src/poetry/repositories/pool.py:90]: relevant: https://github.com/python-poetry/poetry/issues/6670#issuecomment-1264443658

File: src/poetry/repositories/pool.py
Comment (neersighted) [reply src/poetry/repositories/pool.py:27]: Looking at this closer, we do have a lot of shared methods. Let's make a common `AbstractRepository` ABC that both `RepositoryPool` and `Repository` can implement.

File: src/poetry/repositories/pool.py
Comment (neersighted) [reply src/poetry/repositories/pool.py:90]: Given that comment, I think we should normalize names uniformly.

File: src/poetry/repositories/pool.py
Comment (b-kamphorst) [reply src/poetry/repositories/pool.py:27]: It turns out the number of shared methods is quite small: only `find_packages` and `search` overlap (assuming we don't put private methods in the ABC). The method `package` also exists in both, but the signature is different.

Comment (neersighted) [reply src/poetry/repositories/pool.py]: Hmm, we could make `package` common with the use of kwargs. I'm rather inclined to do that because the idea of reasoning about these repos based on ABC instead of concrete type (when possible) is quite attractive. 

Also, I noticed for some reason `CachedRepository` does inherit from ABC without having any `@abstractmethod`s -- we should probably clean that up as well (e.g. add abstract declarations common between implementations).

**Comment by b-kamphorst** [comment]
> I just ran `test_pool.py` with coverage. It seems there is no coverage for `has_primary_repositories`, `find_packages` and `search` and `package` is only tested superficially. It would be nice to add some tests to gain more confidence in the refactoring.

We now have 100% test coverage for `test_repository_pool.py` 😃 

> However, I noticed that without `_lookup` and the Boolean flags there is more iterating in some methods/properties. I don't know how often this code is called during dependency resolution and if this could be a performance bottleneck.

I assumed that we wouldn't do this too often, but it was quite easy to make this constant-time lookup (at the slight expense of clarity).

File: src/poetry/repositories/pool.py
Comment (b-kamphorst) [reply src/poetry/repositories/pool.py]: Done.

File: src/poetry/repositories/pool.py
Comment (b-kamphorst) [reply src/poetry/repositories/pool.py]: I didn't find a way to do this without `# type: ignore[override]`, but perhaps you know a better way?

`CachedRepository` has a single abstract method named `_get_release_info`.

**Comment by b-kamphorst** [comment]
Apparently all tests failed due to an external import by `poetry_plugin_export`. I temporarily fixed it in ea72f31d4cc1ac07f8106a77ee8b7b0c9b75eb4b, but I can imagine you want to revert the whole renaming of `repositories/http.py`.

File: src/poetry/repositories/pool.py
Comment (dimbleby) [reply src/poetry/repositories/pool.py]: in your current draft, only the `AbstractRepository` has kwargs, none of the concrete classes do.  So just remove the kwargs from the abstract class, and also remove all the `# type: ignore[override]` comments?

**Comment by b-kamphorst** [comment]
> Apparently all tests failed due to an external import by `poetry_plugin_export`. I temporarily fixed it in [ea72f31](https://github.com/python-poetry/poetry/commit/ea72f31d4cc1ac07f8106a77ee8b7b0c9b75eb4b) and [9aec915](https://github.com/python-poetry/poetry/commit/9aec91518956c3a3564a89e1eaf517313c1468d3), but I can imagine you want to revert the whole renaming of `repositories/http.py` and `repositories/pool.py`.

This is clearly breaking, I'll revert the changed names tomorrow.

File: src/poetry/repositories/pool.py
Comment (b-kamphorst) [reply src/poetry/repositories/pool.py]: The `RepositoryPool` does, that's the issue..

Comment (dimbleby) [reply src/poetry/repositories/pool.py]: then add kwargs to the `RepositoryPool` only, and not to the abstract base class

Comment (dimbleby) [reply src/poetry/repositories/pool.py]: I re-reviewed the code and stand by plan A: remove the kwargs from the abstract class and remove all of the `# type: ignore[override]` comments.

**Comment by neersighted** [comment]
@b-kamphorst I'm okay with the rename personally -- the plugin is coupled to implementation details and we can work around that with a 3-PR-dance or just some maintainer magic (but we should break the rename out into a separate PR).

Regarding `search`, it's because only PyPI supports search as of now via the deprecated XML-RPC API, and not via PEP 508. Any vestiges of search on non-PyPI repositories is a holdover from when PyPI was not special-cased to use the JSON API: https://warehouse.pypa.io/api-reference/xml-rpc.html

**Comment by radoering** [comment]
I'm strongly for breaking all the renaming out in a separate PR with **renaming only and no other changes**.

> We now have 100% test coverage for `test_repository_pool.py` 😃

Sounds great. 🎉 

> I assumed that we wouldn't do this too often, but it was quite easy to make this constant-time lookup (at the slight expense of clarity).

The only way to be sure is doing some measurements. (On the one hand I don't assume it is called too often or has a significant influence, but on the other hand I can imagine that some of the methods are called for each package lookup.) If others don't mind we can also take the risk.

File: docs/repositories.md
Comment (neersighted) [inline docs/repositories.md:127]: This isn't accurate -- `default` is mentioned twice and the first mention is incorrect. In addition, I would double quote (`"`) "primary" (instead of writing it as `primary`) to make it clear that it's just the internal nickname and not a TOML-level option.

File: src/poetry/console/commands/show.py
Comment (neersighted) [inline src/poetry/console/commands/show.py:201]: Out of scope/for another time, but I am fascinated that we're using `RepositoryPool` here instead of just using `LockedRepository` directly.

File: src/poetry/repositories/repository_pool.py
Comment (neersighted) [inline src/poetry/repositories/repository_pool.py:133]: Instead of special casing here, it's likely better/cleaner to just return an empty list in `LegacyRepository`.

**Review by neersighted — CHANGES_REQUESTED** [review]
Looking a lot more promising overall -- I haven't had time to look at everything exhaustively, but I left some feedback.

On the rename, I suggest we either to that in a separate PR, or you raise some DeprecationWarnings when the old names are imported.

I'd also like to see the kwargs change -- @dimbleby described it accurately, and it would remove the need for `type: ignore[override]` while better encapsulating how these classes are used.

**Comment by radoering** [comment]
> On the rename, I suggest we either to that in a separate PR, or you raise some DeprecationWarnings when the old names are imported.

When renaming you have to cut the commits carefully in order to avoid "breaking" git history. We shouldn't just squash the commits. See python-poetry/poetry-core#482 for example. That's why I suggest doing it in a separate PR.

File: docs/repositories.md
Comment (radoering) [reply docs/repositories.md:127]: It might be inaccurate, but I don't think the first mention is incorrect. IIUC, the second mention is not always correct.

According to https://github.com/python-poetry/poetry/blob/23c4f9b34bf897948b85b44e351d40d064f7efb5/src/poetry/factory.py#L144-L153

- If there is an explicit default, PyPI is not used.
- If there is no explicit default
  - and you don't have any primary repositories, PyPI is the default.
  - and you have primary repositories there is no default. PyPI is **not** the default, but the  secondary with the lowest priority.

The last case is not described properly.

Comment (neersighted) [reply docs/repositories.md:127]: I think we're both agreeing but in different ways... Here is how I would describe the resolution logic:

* First, "primary" sources are considered in the order they are defined. These are entries in the pyproject.toml without `default = true` or `secondary = true`. This is equivalent to pip's `--extra-index-url`.
* Second, the `default` source is considered. If the default source is not overidden in pyproject.toml, this is PyPI. This is equivalent to pip's `--index-url`.
* Finally, all `secondary` sources are considered in the order they are defined. This is a Poetry invention.

Comment (radoering) [reply docs/repositories.md:127]: Not entirely correct: If there is a default source (explicit or implicit), the default source is always the first primary source. (The default source is always considered first.)

Comment (neersighted) [reply docs/repositories.md:127]: I don't disagree on a source code level, but I am referring to the practical behavior an end user sees:

Given a primary source, `click` and friends will be locked from the primary source:
```toml
[tool.poetry.dependencies]
click = "^8.1.3"

[[tool.poetry.source]]
name = "piwheels"
url = "https://www.piwheels.org/simple"
```

Given a secondary source, `click` and friends will be locked from the default source (PyPI):
```toml
[tool.poetry.dependencies]
click = "^8.1.3"

[[tool.poetry.source]]
name = "piwheels"
url = "https://www.piwheels.org/simple"
secondary = true
```

Skimming the current code, I'm actually a bit confused on how this is implemented, but it is what we currently do.

Comment (neersighted) [reply docs/repositories.md:127]: Indeed, salting some prints into `add_repository` yields the following: 
```
piwheels 0
pypi 1
[<poetry.repositories.legacy_repository.LegacyRepository object at 0x7fdebc55dae0>, <poetry.repositories.pypi_repository.PyPiRepository object at 0x7fdebc5be3e0>]
```

I do think that the logic is tied into a hard to decipher pretzel, and the `RepositoryPriorities` enum in this PR needs to change.

Comment (radoering) [reply docs/repositories.md:127]: In a discussion on Discord we came to the conclusion that

> * First, "primary" sources are considered in the order they are defined. These are entries in the pyproject.toml without `default = true` or `secondary = true`. This is equivalent to pip's `--extra-index-url`.
> 
> * Second, the `default` source is considered. If the default source is not overidden in pyproject.toml, this is PyPI. This is equivalent to pip's `--index-url`.
> 
> * Finally, all `secondary` sources are considered in the order they are defined. This is a Poetry invention.

describes how it should be. However, it isn't the current behavior. So let's drop the documentation for now, do only the refactoring here and add documentation + bugfix to comply with the description above in a follow-up PR because it's easier to do after the refactoring than before.

Sorry for the back and forth. We really appreciate your contribution. 🙏

File: src/poetry/repositories/pool.py
Comment (b-kamphorst) [reply src/poetry/repositories/pool.py]: I've processed the change. I guess it is a matter of preference -- as I user I like to be informed of potential deviations from the ABC (API to me), but it did get quite ugly for little benefit.

File: docs/repositories.md
Comment (b-kamphorst) [reply docs/repositories.md:127]: Ha don't worry about it! It took me quite some time to get familiar with enough code to work on this and the PyPI insertion and default/primary/secondary sources got me confused -- glad to see I'm not the only one. I'd be more than happy to change the behavior into something that can be explained easily (but I figured that a pure refactor should not change the current behavior). I'll take this up in my attempt at https://github.com/python-poetry/poetry/pull/5984.

File: src/poetry/repositories/repository_pool.py
Comment (b-kamphorst) [reply src/poetry/repositories/repository_pool.py]: Fixed. We might even clean up `LegacyRepository` further by removing all references to `self._packages` -- they seem to never get populated.

**Comment by b-kamphorst** [comment]
It appears that master (44a89cbd7530bc3c50c87f672c221f61125faf54) breaks all tests?

**Comment by neersighted** [comment]
> It appears that master ([44a89cb](https://github.com/python-poetry/poetry/commit/44a89cbd7530bc3c50c87f672c221f61125faf54)) breaks all tests?

You've got a typing issue.

File: src/poetry/repositories/pool.py
Comment (neersighted) [reply src/poetry/repositories/pool.py]: Alternatively, we could just ignore `kwargs` on every implementation except RepositoryPool, if that makes you feel better :laughing:

**Comment by b-kamphorst** [comment]
> > It appears that master ([44a89cb](https://github.com/python-poetry/poetry/commit/44a89cbd7530bc3c50c87f672c221f61125faf54)) breaks all tests?
> 
> You've got a typing issue.

Fixed that. Locally the tests also failed in Sint places but apparently that is resolved... 

I prepared another PR with the name changes. I'll present it when this PR is closed.

Thanks for your time, I enjoy the collaboration!

**Comment by b-kamphorst** [comment]
Hi @radoering, @neersighted, are there any other changes that need to be processed here (at some point @neersighted mentioned that the `RepositoryPriorities` needs to change, but that discussion was dropped for this PR) or can the PR be merged?

File: src/poetry/repositories/legacy_repository.py
Comment (radoering) [inline src/poetry/repositories/legacy_repository.py:38]: Why that?

File: src/poetry/repositories/abstract_repository.py
Comment (radoering) [inline src/poetry/repositories/abstract_repository.py:11]:
```suggestion
    from poetry.core.constraints.version import Version
```

File: src/poetry/repositories/legacy_repository.py
Comment (radoering) [reply src/poetry/repositories/legacy_repository.py:38]: Can we add a comment in the code why we are just returning an empty list please?

File: src/poetry/repositories/pool.py
Comment (neersighted) [inline src/poetry/repositories/pool.py:21]: I like the introduction of the dataclass -- with that present, I think this can be named cleaner as `Priority`.

File: src/poetry/repositories/pool.py
Comment (neersighted) [inline src/poetry/repositories/pool.py:29]: I think we should either make this `frozen=true` or `NamedTuple` as there is no reason I see to leave it mutable.

File: src/poetry/repositories/pool.py
Comment (neersighted) [inline src/poetry/repositories/pool.py:35]: It looks like the name got reverted to `Pool`. Sorry if I was unclear before -- I'd like to keep the name `RepositoryPool` in this PR, even if we do the file-level renames in another.

File: src/poetry/repositories/pool.py
Comment (neersighted) [inline src/poetry/repositories/pool.py:45]: This is pretty messy -- it doesn't look like we ever call `Pool(None)`, so it should be perfectly safe to change this to a default value without any other code changes.

File: src/poetry/repositories/pool.py
Comment (neersighted) [inline src/poetry/repositories/pool.py:69]: We could do this in another PR, but in the long term I'd like to add a `RepositoryName` type that represents a normalized name, and centralize all normalization code there.

File: src/poetry/repositories/pool.py
Comment (neersighted) [inline src/poetry/repositories/pool.py:73]: I'd rather see this use `has_repository()`.

File: src/poetry/repositories/pool.py
Comment (neersighted) [inline src/poetry/repositories/pool.py:84]: Same as above, I'd rather see this use `has_repository()`.

File: src/poetry/repositories/pool.py
Comment (neersighted) [inline src/poetry/repositories/pool.py:98]: I'd rather use `priority` here and hoist this out of the else to be on top.

File: src/poetry/repositories/pool.py
Comment (neersighted) [inline src/poetry/repositories/pool.py:105]: The old function had error checking -- this can now throw an `IndexError` -- we should probably check for existence first since we silently swallowed errors before.

**Review by neersighted — CHANGES_REQUESTED** [review]
Overall looks very promising. I have some concerns about the specific implementation, but otherwise the new tests look good and the architecture seems sound. Good work cleaning this up!

File: src/poetry/repositories/legacy_repository.py
Comment (neersighted) [reply src/poetry/repositories/legacy_repository.py:38]: Or a docstring :)

File: src/poetry/repositories/pool.py
Comment (neersighted) [reply src/poetry/repositories/pool.py:105]: (and we should also have a test for deleting a non-extant repository)

File: src/poetry/repositories/legacy_repository.py
Comment (b-kamphorst) [reply src/poetry/repositories/legacy_repository.py:38]: Looking into this again, I wonder why we return an empty list indeed. The rationale for changing this is to remove the special casing in `Pool.search`. However, `LegacyRepository.search` only browses `LegacyRepository.packages` and inspects their names (one may debate whether that is 'searching', but it is what happens). As such, it appears that *even if `LegacyRepository.packages` is populated*, this should not be a heavy task. Should we undo the change to `LegacyRepository`? If not, what is the rationale to be provided in the docstring?

Related (though a bit of a sidestep): it appears that the `search` command (console/commands/search) considers `pypi` only and ignores all other repositories (even if `pypi` is disabled). Is that the intended behaviour? It seems odd.

File: src/poetry/repositories/pool.py
Comment (b-kamphorst) [reply src/poetry/repositories/pool.py:35]: That will still break [the plugin test code](https://github.com/python-poetry/poetry-plugin-export/blob/main/tests/conftest.py#L16), so I think it is best to combine the file+ object renaming in a single PR. Do you agree?

File: src/poetry/repositories/pool.py
Comment (b-kamphorst) [reply src/poetry/repositories/pool.py:45]: We do call `Pool()` in many places. Which default value are you thinking of (assuming you don't want to have a mutable object as default)?

File: src/poetry/repositories/pool.py
Comment (b-kamphorst) [reply src/poetry/repositories/pool.py]: See [278c432](https://github.com/python-poetry/poetry/pull/6669/commits/278c4325e3bb6a4a11e220fd18aab83217583b44) for an attempt. I'm not too happy with it, but let me know what you think.

Comment (b-kamphorst) [reply src/poetry/repositories/pool.py]: Ah again this breaks plugin...

**Comment by b-kamphorst** [comment]
> Once review is addressed, I'd like to see cleaner history -- it would be nice to introduce the new tests and then the refactor (assuming the tests pass without the refactor -- I didn't look closely to see if they would run against the old implementation).

This took me a lot of time, but at least I learned a bit more about git (plenty to learn left). The added tests are now in the first commit (not 100% at that commit just yet).

**Comment by neersighted** [comment]
I personally would drop the `RepositoryName` commits from this PR and move them to a subsequent PR -- I think that we might want to think a little more about the design/implementation and it clutters this one too much.

File: src/poetry/repositories/pool.py
Comment (neersighted) [reply src/poetry/repositories/pool.py:36]: Ah, I missed that -- fair enough!

File: src/poetry/repositories/pool.py
Comment (neersighted) [reply src/poetry/repositories/pool.py:48]: Ah, shoot, I apparently have spent too much time in other languages (*cough*Ruby*cough*) -- I appear to have forgotten about mutable default values last night. My bad :laughing:

File: src/poetry/repositories/pool.py
Comment (neersighted) [inline src/poetry/repositories/pool.py:48]: I'd rather write this as

```python
if repositories is None:
    repositories = []
```

It's more correct as if we use a boolean evaluation, an empty list is false-y. Using the caller's list isn't necessary here, but it doesn't hurt to use the more correct form.

File: src/poetry/repositories/legacy_repository.py
Comment (neersighted) [reply src/poetry/repositories/legacy_repository.py:38]: It's indeed a good question. I believe currently this is because `poetry search` only ever historically searched PyPI, and when we refactored repositories the first time (the old code was quite jumbled), `search` was implemented on the base `Repository` class even if it doesn't make a ton of sense on `LegacyRepository` as we only populate it on demand.

If we teach the search command to use `LegacyRepository` and use some sort of lightweight `PackageVersion` structure that doesn't require inspection like creating a `Package`, this is probably worth revisiting (see https://github.com/python-poetry/poetry/issues/2446) -- but that's work for another day. For now a regular comment instead of docstring explaining we're stubbing this out to avoid changing behavior/it's a TODO to rethink search so all source types can support it likely makes more sense.

**Comment by neersighted** [comment]
@b-kamphorst I've pushed up a rebase of your branch that squashes the changes down into mergable commits -- I think this is good to go once we address the final review commits, as long as we keep the history clean.

File: src/poetry/repositories/pool.py
Comment (neersighted) [inline src/poetry/repositories/pool.py:66]:
```suggestion
            prio_repo.priority is priority for prio_repo in self._repositories.values()
```

Enums should be compared using identity, as per https://docs.python.org/3/library/enum.html#comparisons

File: src/poetry/repositories/pool.py
Comment (neersighted) [inline src/poetry/repositories/pool.py:26]:
```suggestion
    DEFAULT = enum.auto()
    PRIMARY = enum.auto()
    SECONDARY = enum.auto()
```

Very minor nit, but I prefer `enum.auto()` here as `auto()` is a very generic function name.

File: src/poetry/repositories/pool.py
Comment (neersighted) [inline src/poetry/repositories/pool.py:52]: We're creating a list here when the view object will work just fine with `sorted()`.

File: src/poetry/repositories/pool.py
Comment (neersighted) [inline src/poetry/repositories/pool.py:41]:
```suggestion
        super().__init__("poetry-pool")
```

Let's keep this consistent with the class name -- it can change back in the rename PR.

**Review by neersighted — CHANGES_REQUESTED** [review]
Final batch of requested changes -- should be good to go after we discuss/address these.

File: src/poetry/repositories/pool.py
Comment (neersighted) [inline src/poetry/repositories/pool.py:76]: This should probably be an `IndexError` instead.

File: src/poetry/repositories/pool.py
Comment (neersighted) [inline src/poetry/repositories/pool.py:92]:
```suggestion
            raise ValueError("Only one repository can be the default.")
```

File: src/poetry/repositories/pool.py
Comment (neersighted) [inline src/poetry/repositories/pool.py:88]:
```suggestion
                "A repository with name {repository_name}"
                f" was already added."
```

I think we can drop the "case-independent" bit as we're printing the name we're using in the exception.

**Comment by b-kamphorst** [comment]
> @b-kamphorst I've pushed up a rebase of your branch that squashes the changes down into mergable commits -- I think this is good to go once we address the final review comments, as long as we keep the history clean.

I have processed the suggested changes (in a separate commit for your convenience). If they are in line with your expectations, feel free to squash the final commit into another commit and make this fly :)

**Review by neersighted — APPROVED** [review]
LGTM. FWIW, I would suggest getting familiar with interactive rebases and how to leverage the `edit` rebase command (as well as the `git reset` CLI command) to manipulate the history of branches. When you're doing larger, more ambitious PRs like this, you generally want to take the approach you took and make the history as atomic/easy to follow for reviewers as possible.

Then, when it's time for merge, you'll want to squash your commits down. In the simple/smaller PR case a squash of everything is just fine. But in a larger PR like this it's helpful to capture some of the evolutionary steps as commits in history -- which is when you want to master that process of munging history during an interactive rebase.

**Pull request merged**
